File: MATTHEW/leetcode/AIcoding/3.DeepSeek-v3/java_solutions/1031-maximum-sum-of-two-non-overlapping-subarrays.java
class Solution {
    public int maxSumTwoNoOverlap(int[] nums, int firstLen, int secondLen) {
        int[] prefixSum = new int[nums.length + 1];
        for (int i = 0; i < nums.length; i++) {
            prefixSum[i + 1] = prefixSum[i] + nums[i];
        }
        return Math.max(maxSum(prefixSum, firstLen, secondLen), maxSum(prefixSum, secondLen, firstLen));
    }

    private int maxSum(int[] prefixSum, int L, int M) {
        int maxL = 0;
        int res = 0;
        for (int i = L + M; i < prefixSum.length; i++) {
            maxL = Math.max(maxL, prefixSum[i - M] - prefixSum[i - M - L]);
            res = Math.max(res, maxL + prefixSum[i] - prefixSum[i - M]);
        }
        return res;
    }
}

File: leetcode/AIcoding/2.Gemini-2.0/cpp_solutions/1031-maximum-sum-of-two-non-overlapping-subarrays.cpp
#include <vector>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxSumTwoNoOverlap(vector<int>& nums, int firstLen, int secondLen) {
        int n = nums.size();
        vector<int> prefixSum(n + 1, 0);
        for (int i = 1; i <= n; ++i) {
            prefixSum[i] = prefixSum[i - 1] + nums[i - 1];
        }

        int maxSum = 0;
        for (int i = firstLen; i <= n; ++i) {
            int firstSubarraySum = prefixSum[i] - prefixSum[i - firstLen];
            for (int j = secondLen; j <= i - firstLen; ++j) {
                int secondSubarraySum = prefixSum[j] - prefixSum[j - secondLen];
                maxSum = max(maxSum, firstSubarraySum + secondSubarraySum);
            }
            for (int j = i + firstLen; j <= n; ++j) {
                int secondSubarraySum = prefixSum[j] - prefixSum[j - secondLen];
                maxSum = max(maxSum, firstSubarraySum + secondSubarraySum);
            }
        }
        return maxSum;
    }
};